### PR TITLE
Modernize JavaScript on integrations page

### PR DESCRIPTION
**Pull request by Daniel15**
## Proposed change

This JavaScript code looks pretty old, likely written when Internet Explorer support was still needed. These days, most people (especially those using Home Assistant) would be using modern browsers and the code can be simplified a bit.

This diff modernizes the JavaScript on the integrations page:
 - Removes third-party libraries:
   - jQuery: 72.4 KB min, 25.2 KB min + gzip.
   - Mustache: 9.5KB min, 3.6 KB min + gzip.
   - Lazyload: 5.1 KB min, 3.84 KB min + gzip. Turns out this page is already using the browser's native lazy loading functionality by setting `loading="lazy"` on the images.
 - Removes some magic numbers (e.g. `.substring(8)` and `.substring(9)`).
 - Fixes a bug where any keyup events in the search field causes the integrations to re-render
 - Cleans up code for building image URLs.
 - Uses function arrow syntax for callbacks.
 - Uses template strings instead of Mustache template.
 - Uses `history.pushState` directly since all browsers support it now.
 - Strips debounce function to bare minimum implementation.
 - Renames some variables so that the names are consistent (camel case).

## Test plan

Page shows featured integrations by default:
image

Clicking a category shows only integrations in that category. URL correctly updates to e.g. `/#binary-switch`:
image

Clicking "all" shows all, with URL of `/#all`:
image

Selecting version from dropdown only shows integrations added in that version, with URL of `/#version/2023.12`:
image

Typing in search box does a search after 500ms, with a URL of `/#search/mqtt`:
image

## Type of change
- [ ] Spelling, grammar or other readability improvements (`current` branch).
- [ ] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logos and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information
N/A

## Checklist
N/A

## Summary by CodeRabbit

- **New Features**
  - Enhanced search functionality for better component filtering.
  - Improved filtering based on search terms, versions, and categories.
  - Dynamic rendering of components based on selected filters.

- **Improvements**
  - Optimized event handling and input sanitization for smoother user experience.
  - Replaced jQuery with modern JavaScript methods for better performance.

### CODE REVIEW

File: source/integrations/index.html [hunk 1]
```diff
@@ -131,185 +105,169 @@
...
+
+    const searchInputEl = document.querySelector('.component-search input');
+    const componentContainerEl = document.querySelector('#componentContainer');
 
     function init() {
       // do the lowerCase transformation once
-      for (i = 0; i < allComponents.length; i++) {
-        title = allComponents[i].title.toLowerCase();
-        domain = allComponents[i].domain;
-        title_normalized = title
```
Comment (Daniel15) [inline source/integrations/index.html:142]: Other variables were camelcase - renamed these to match.

File: source/integrations/index.html [hunk 2]
```diff
@@ -131,185 +105,169 @@
...
       applyFilter();
-
-      return false;
     });
 
     /**
-     * Simple debounce implementation, based on http://davidwalsh.name/javascript-debounce-function
+     * Simple debounce implementation
      */
-    function debounce(func, wait, immediate) {
```
Comment (Daniel15) [inline source/integrations/index.html:287]: For the use case on this page:
 - `immediate` was never `true` - deleted.
 - `func` never has any arguments - removed arg handling

**Comment by frenck** [comment]
I've rebased the PR on top of the latest `current` version, and it seems to break the current site:

CleanShot 2024-06-08 at 10 16 52@2x

**Comment by Daniel15** [comment]
@frenck Thanks for letting me know. I'll take a look and update the PR, hopefully this weekend.

**Comment by frenck** [comment]
The deploy preview shows the issue as well:

<https://deploy-preview-30964--home-assistant-docs.netlify.app/integrations/>

**Comment by Daniel15** [comment]
@frenck This should be good to go now. It looks like there was a merge issue that caused the problem you saw.

**Review by frenck — APPROVED** [review]
Nice! Thanks, @Daniel15 👍 

../Frenck

**Pull request merged**